feat(search): add search departments, lists, candidates

drendog opened the pull request:
# TODO

- [x] solve conflicts
- [x] search by lists. Closes #66 
- [x] search by candidates. Closes #56 
- [x] refactor and replace for loops with map
- [x] improve graphics

Live version [here](https://drendog.github.io/UNICT-Elezioni/)
any advice is welcome.

Helias commented on src/app/App.tsx:
```diff
@@ -3,22 +3,23 @@ import {
   Route,
   HashRouter,
-  Redirect,
+  Redirect
```
I prefer to leave the comma when it is possible

Helias commented on src/app/App.tsx:
```diff
@@ -3,22 +3,23 @@ import {
   <div className="App">
     <HashRouter basename="/">
       <Menu />
-      <br />
```
we should avoid to use br, so, good job :+1:

Helias commented on src/app/App.tsx:
```diff
@@ -55,9 +56,9 @@ const App: FunctionComponent = () => (
-              years.map((y) => <Results anno={y} path={`dipartimenti/${d}`} details={false} showDetailsList />),
+              <hr className="my-5" key={`hr${d}`} />,
+              <h3 key={`h3${d}`}><a href={`#/dipartimento/${d}`}>{d.replace(/_/g, ' ')}</a></h3>,
+              years.map((y) => <Results anno={y} path={`dipartimenti/${d}`} details={false} key={`${d}${y}`} showDetailsList />)
```
`showDetailsList` is not used I think, so you could even remove it and put in the component props it as a "optional" attribute using `?` at the end of the attribute in the Prop interface

drendog replied:
I use `showDetailsList` for popover that show the seats table, because if I use this popover on the Senate, the table will be very long, and it can get out of the page limits.

Helias commented on src/app/results/DetailList/DetailsList.tsx:
```diff
@@ -28,26 +27,26 @@ export const DetailsList = (props: Props) => {
   }
 
-  function generateVotesRow(): JSX.Element {
+  function generateVotesRow (): JSX.Element {
```
why that space? :imp:

Helias commented on src/app/results/DetailList/DetailsList.tsx:
```diff
@@ -62,7 +61,7 @@ export const DetailsList = (props: Props) => {
-  )
-}
+  );
+};
```
I don't think that this last semi-colon is needed

Helias commented on src/app/results/ListLogo/ListLogo.tsx:
```diff
@@ -23,8 +22,7 @@ export const ListLogo = (props: Props) => {
-
-}
+  );
+};
```
same

Helias commented on src/app/search/SearchEngine.ts:
```diff
@@ -0,0 +1,28 @@
+    this._data = data;
+  }
+
+  search (str: string): string[] {
```
`function()` not `function ()` :(

Helias commented on src/index.tsx:
```diff
@@ -6,7 +6,7 @@ import * as serviceWorker from './serviceWorker';
 ReactDOM.render(
   <App />,
-  document.getElementById('root'),
+  document.getElementById('root')
```
I would leave the comma

Helias commented on src/app/search/SearchForm.tsx:
```diff
@@ -0,0 +1,42 @@
+    <ListGroup className={formValue.length > 0 ? 'suggestions-list' : 'd-none' }>
+      {generateSuggestions()}
+    </ListGroup>
+  </div>);
```
this feature is fantastic

Helias commented on src/app/search/SearchForm.tsx:
```diff
@@ -0,0 +1,42 @@
+  }
+
+  function generateSuggestions (): JSX.Element[] {
+    const results = [];
+    for (const suggestion of suggestions) {
+      results.push(
+        <a href={`#/dipartimento/${suggestion}`}>
+          <ListGroup.Item action variant="light">
+            {suggestion.replaceAll('_', ' ')}
+          </ListGroup.Item>
+        </a>
+
+      );
+    }
+    return results;
```
```suggestion
    return suggestions.map((s) => {
        <a href={`#/dipartimento/${s}`}>
          <ListGroup.Item action variant="light">
            {s.replaceAll('_', ' ')}
          </ListGroup.Item>
        </a>
    });
```
```

drendog commented:
so, about the spaces after the function name and the comma style, this is a rule defined by [standardjs](https://standardjs.com/rules.html) `space-before-function-paren` and `comma-style`, eslint will warn us for the missing space, but if you don't like it, I can disable this rule at any time. About the semi-colon, I added the rule (semi) for forcing its consistent use, and I can disable this too.

Helias commented:
@drendog I think that probably we need to use tslint, anyways, personally I hate those spaces 😅

drendog commented:
Added the search by list, it is not ended yet, but any advice is welcome.
[TSLint](https://palantir.github.io/tslint/) has been **deprecated** as of 2019, and their devs advice us to migrate to ESLint. It is customizable and seems to work fine, if you don't like some style, you can add some rules on '.eslintrc.json'. Now I added the rule for remove those spaces, if the comma style looks bad too, I can remove it if you want.

Helias commented:
@drendog 😨, ok 😔

Helias commented on src/app/results/DetailList/Seggi.ts:
```diff
@@ -29,16 +29,16 @@ class OtherSeggi implements ISeggi {
   }
 
-  getSeggi (): string[] {
+  getSeggi(): string[] {
```
```suggestion
  get seggi(): string[] {
```

Helias commented on src/app/results/Results.tsx:
```diff
@@ -25,9 +24,9 @@ export const Results = (props: Props): JSX.Element => {
-        </Table>
+        {props.details &&
+          <DetailsTable seggio={props.seggio} data={data} anno={props.anno} />
+        }
```
very very nice!

Helias commented on src/app/results/ResultsTable.tsx:
```diff
@@ -42,51 +40,53 @@ export const ResultTable = (props: Props): JSX.Element => {
       tableRows.push(
         <tr key={`${props.anno}-${i}`}>
-          {Object.keys(results).map((l) => !!l &&
+          {Object.keys(results).map((l: string) => l !== 'undefined' &&
```
I think that this can be improved

Helias commented on src/app/results/ResultsTable.tsx:
```diff
@@ -132,7 +132,7 @@ export const ResultTable = (props: Props): JSX.Element => {
   );
 
-  function toggleBody (e: any): void {
+  function toggleBody(e: any): void {
```
:relieved:

Helias commented on src/app/search/SearchEngine.ts:
```diff
@@ -0,0 +1,190 @@
+      }
+    }
+    return true;
+  }
```
I don't really like the logic here, even because when you do `isPresent = true;` in the next iteration you could not find the namePart and set it false back.
You should add a `break;` after `isPresent = true;`

But you can improve this in a really better way.
- remove the variable isPresent
- if you find it `return true` directly
- otherwise `return false`

Helias replied:
nevermind, that's ok

Helias commented on src/app/search/SearchEngine.ts:
```diff
@@ -0,0 +1,190 @@
+  private maxResults = 8;
+  private limits = new SearchLimits(this.maxResults);
+
+  private constructor() {
```
:thinking:

Helias replied:
ah ma e' un singleton :grin:

Helias commented on src/app/search/SearchForm.tsx:
```diff
@@ -0,0 +1,88 @@
+          {suggestion.replaceAll('_', ' ')}
+        </ListGroup.Item>
+      </a>
+    ) as any;
```
sooner or later we need to add the types for everything

Helias commented on src/app/search/SearchForm.tsx:
```diff
@@ -0,0 +1,88 @@
+      value={formValue}
+      onChange={onInputFormChange}
+      placeholder="Cerca dipartimento, candidato, lista..." />
+    <ListGroup className={formValue.length > 0 ? 'suggestions-list' : 'd-none'}>
```
you can even omit the `> 0`

Helias commented on src/data/entities.ts:
```diff
@@ -0,0 +1,19 @@
+  Nucleo_di_valutazione: '#/ndv',
+  Consiglio_di_amministrazione: '#/cda',
+  Comitato_per_lo_sport_universitario: '#/csu'
+};
```
may an `enum` can be a better choice (?)

drendog replied:
maybe no, because i'm using `IEntitiesPath` interface

Helias commented on src/data/entities.ts:
```diff
@@ -0,0 +1,19 @@
+  Comitato_per_lo_sport_universitario: '#/csu'
+};
+
+export default entities;
```
you could directly `export entities` instead of using `export default entitites;`

Helias commented on src/app/App.tsx:
```diff
@@ -3,64 +3,70 @@ import {
           </Route>
           <Route exact path="/contatti">
-            <Contacts></Contacts>
+            <Contacts/>
```
:+1:

drendog commented:
approved?

Helias commented:
yes
